src/main/java/org/jenkinsci/main/modules/sshd/PublicKeyAuthenticatorImpl.java

```diff
@@ -20,27 +28,66 @@
     private final PublicKeySignatureWriter signatureWriter = new PublicKeySignatureWriter();
 
     public boolean authenticate(String username, PublicKey key, ServerSession session) {
-        LOGGER.fine("Authentication attempted from "+username+" with "+key);
-        User u = User.get(username, false);
-        if (u==null) {
-            LOGGER.fine("No such user exists: "+username);
+        User user = this.retrieveOnlyKeyValidatedUser(username, key);
+
+        if (user == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
-        UserPropertyImpl sshKey = u.getProperty(UserPropertyImpl.class);
-        if (sshKey==null) {
-            LOGGER.fine("No SSH key registered for user: "+username);
+        Authentication auth = this.verifyUserUsingSecurityRealm(user);
+        if (auth == null) {
+            SecurityListener.fireFailedToAuthenticate(username);
             return false;
         }
 
+        UserDetails userDetails = new SSHUserDetails(username, auth);
+        SecurityListener.fireAuthenticated(userDetails);
+        return true;
+    }
+
+    private @CheckForNull User retrieveOnlyKeyValidatedUser(String username, PublicKey key) {
+        LOGGER.fine("Authentication attempted from " + username + " with " + key);
+        User u = User.get(username, false, Collections.emptyMap());
+        if (u == null) {
+            LOGGER.fine("No such user exists: " + username);
+            return null;
+        }
+
+        UserPropertyImpl sshKey = u.getProperty(UserPropertyImpl.class);
+        if (sshKey == null) {
+            LOGGER.fine("No SSH key registered for user: " + username);
+            return null;
+        }
+
         String signature = signatureWriter.asString(key);
         if (!sshKey.isAuthorizedKey(signature)) {
-            LOGGER.fine("Key signature didn't match for the user: "+username+" : " + signature);
-            return false;
+            LOGGER.fine("Key signature didn't match for the user: " + username + " : " + signature);
+            return null;
         }
 
-        return true;
+        return u;
+    }
+
+    private @CheckForNull Authentication verifyUserUsingSecurityRealm(@Nonnull User user) {
+        try {
+            return user.impersonate();
+        } catch (UsernameNotFoundException e) {
+            LOGGER.log(Level.FINE, user.getId() + " is not a real user accordingly to SecurityRealm", e);
+            return null;
+        }
     }
 
     private static final Logger LOGGER = Logger.getLogger(PublicKeyAuthenticatorImpl.class.getName());
+
+    public static class SSHUserDetails extends org.acegisecurity.userdetails.User {
```

Comment:
This inner class respects the hashCode / equals method of the inherited class User. The goal of the inheritance is to give listener of the Security event the possibility to know where the UserDetails come from.